### ISSUE TITLE
SIGPIPE handling is wrong

### ISSUE THREAD

**Issue by treyharris** [rakudo/rakudo#3004]
## The Problem
The current handling of SIGPIPE is wrong. “Wrong”, in that its exposure to the language is effectively useless. Rakudo’s  automatic behavior handling the signal when `SIGPIPE` goes unmentioned in user code has a potentially better failure mode in certain circumstances than Perl 5’s, but that default behavior is forced on you in all circumstances, which is what is wrong.

## Expected Behavior
```
% # Reimplementation of Unix “yes” command
% perl6 -e 'loop { say "y" }' | head
y
y
y
y
y
y
y
y
y
y
%
```

## Actual Behavior
```
% # Reimplementation of Unix “yes” command
% perl6 -e 'loop { say "y" }' | head
y
y
y
y
y
y
y
y
y
y
Failed to write bytes to filehandle: Broken pipe
  in block <unit> at -e line 1

Unhandled exception: Failed to write bytes to filehandle: Broken pipe
   at SETTING::src/core/Exception.pm6:485  (/home/trey/perl6/rakudo-star/rakudo-star-2019.03/install/share/perl6/runtime/CORE.setting.moarvm:<anon>)
 from gen/moar/stage2/NQPHLL.nqp:1854  (/home/trey/perl6/rakudo-star/rakudo-star-2019.03/install/share/nqp/lib/NQPHLL.moarvm:command_eval)
 from src/Perl6/Compiler.nqp:38  (/home/trey/perl6/rakudo-star/rakudo-star-2019.03/install/share/nqp/lib/Perl6/Compiler.moarvm:command_eval)
 from gen/moar/stage2/NQPHLL.nqp:1773  (/home/trey/perl6/rakudo-star/rakudo-star-2019.03/install/share/nqp/lib/NQPHLL.moarvm:command_line)
 from gen/moar/main.nqp:48  (/home/trey/perl6/rakudo-star/rakudo-star-2019.03/install/share/perl6/runtime/perl6.moarvm:MAIN)
 from gen/moar/main.nqp:1  (/home/trey/perl6/rakudo-star/rakudo-star-2019.03/install/share/perl6/runtime/perl6.moarvm:<mainline>)
 from <unknown>:1  (/home/trey/perl6/rakudo-star/rakudo-star-2019.03/install/share/perl6/runtime/perl6.moarvm:<main>)
 from <unknown>:1  (/home/trey/perl6/rakudo-star/rakudo-star-2019.03/install/share/perl6/runtime/perl6.moarvm:<entry>)
```

## Consideration

The default behavior of Unix (and picked up by Perl 5) was that SIGPIPE was trappable, but if untrapped _or ignored on standard output or error only_ would result in the process being killed and a SIGPIPE being sent to the downstream process in turn.

This made sense when streaming textual I/O via STDIN/STDOUT/STDERR was the great majority of I/O done with Perl in the field; the most common way for a consumer or producer to indicate it was done was simply to exit, resulting in a downstream SIGPIPE. (SIGPIPE is never sent upstream.)

Rakudo’s behavior—assuming it’s intentional—makes somewhat better sense for output to file descriptors _other_ than `$*OUT` and `$*ERR`: handle the signal and turn it into an `X::AdHoc` exception. If the pipe was managed by a dispatcher of some sort, you may need to take further action, you may not, but either way, you could do it in a `CATCH` rather than a signal handler. If you persist in attempting to use the IO object after getting the SIGPIPE, that’s on you and you should expect instability.

The issue is that, in Unix standard I/O, output will normally be buffered but autoflushing, so when you receive the SIGPIPE, even if you try to handle it, you will almost inevitably receive another while doing so—so you die with the `Broken pipe` backtrace anyway.

Perl 5’s default behavior is probably not right for a modern network-oriented language. (Most network-oriented daemon code in Perl 5 ignores SIGPIPE.) But ignoring the line-oriented text-munging use case entirely seems like a bigger break from Perl 5 than we should be making—we want the Easy Things to _still_ Be Easy.

To look at another network-oriented language frequently used by system programmers, Go’s behavior here, as described in [its `signal` document](https://golang.org/pkg/os/signal/#hdr-SIGPIPE), seems quite correct for Perl 6 (once translated into Perl 6 terms):
> When a Go program writes to a broken pipe, the kernel will raise a SIGPIPE signal.
>
> If the program has not called Notify to receive SIGPIPE signals, then the behavior depends on the file descriptor number. A write to a broken pipe on file descriptors 1 or 2 (standard output or standard error) will cause the program to exit with a SIGPIPE signal. A write to a broken pipe on some other file descriptor will take no action on the SIGPIPE signal, and the write will fail with an EPIPE error.
>
> If the program has called Notify to receive SIGPIPE signals, the file descriptor number does not matter. The SIGPIPE signal will be delivered to the Notify channel, and the write will fail with an EPIPE error.
>
> This means that, by default, command line programs will behave like typical Unix command line programs, while other programs will not crash with SIGPIPE when writing to a closed network connection.

It looks as if nqp can’t currently achieve this. One can use `NativeCall` to implement die or ignore behavior (see the code sample from @Leont in  #2695 for how), but not actual handling.

The use cases for actual handling of SIGPIPE on `$*OUT` or `$*ERR` are limited, especially without being able to determine which file descriptor SIGPIPE arrived from. If you _could_ determine this (again, not something contemplated by nqp at present), you might decide that if `$*OUT` but not `$*ERR` had SIGPIPE’d, you would want to try to log a message (to a file—sending any message whatsoever to `$*ERR` itself after a SIGPIPE is very bad manners in Unix) and then run finalizations that don’t do standard I/O before exiting.

But as is, setting aside the `NativeCall` hack, there’s literally no way to do anything useful with SIGPIPE on standard output, and there’s no way to achieve the Perl 5 default behavior, and that’s just wrong.

**Comment by treyharris** [comment]
(If you’re reading this issue via email rather than web interface, I got my upstreams and downstreams confused and fixed a couple details after finding the relevant old Unix manual pages, so you may want to read it [there](https://github.com/rakudo/rakudo/issues/3004) instead.)

**Comment by Leont** [comment]
The confusing thing here is that there are two exceptions. One is for the write, and the other is probably because `$*OUT` gets flushed when the program is done.

The first can be caught and handled, the second one can't and that is a problem..

**Comment by treyharris** [comment]
@Leont Agreed, but before getting fancy I suggest we simply restore the Perl 5 default behavior in the least-annoying way—which is the way the snippet of Go documentation above describes.

There’s an interesting [Stack Overflow question](https://stackoverflow.com/questions/108183/how-to-prevent-sigpipes-or-handle-them-properly) question that shows the disconnect here—it goes almost unmentioned that the behavior described is only correct for sockets, not for pipes. If you’re thinking about SIGPIPE as a nuisance that must be dealt with, you deal with it differently than if you take it as a natural alternative way to exit that you probably won’t deal with, or if you deal with, just very carefully.

**Comment by b2gills** [comment]
After writing this: https://gist.github.com/treyharris/489a71a643511895ac272e963eecc973#gistcomment-2949639

And reading this: http://paste.scsys.co.uk/585019
```perl6
use NativeCall;
my sub signal(int32 $signal, int64 $func) is native { * }
signal(+SIGPIPE, 0);

loop {
    say 'y';
}
```

It looks like the way it currently works can be improved with this `END` incantation:
```perl6
END {
  use NativeCall;
  my sub signal(int32 $signal, int64 $func) is native { * }
  signal(+SIGPIPE, 0);
}

sub MAIN(Str $output="y") {

  loop {
    # A:
    # last unless try say $output

    # B:
    CATCH { default { last } }
    say $output;
  }
  note 'done'; # will get printed to $*ERR
}
```

---

Basically ignore pipe signals while the program is exiting.  
(If we are exiting, we probably don't care about any further pipe errors.)

This seems like a relatively uncontentious change.

Treating `$*OUT` and `$*ERR` differently can then be discussed later.

**Comment by Leont** [comment]
That is an early version of my nativecall snipplet, the `int64` should probably be a `long` if you want it to work on 32-bit architectures.

**Comment by Leont** [comment]
But yeah, setting the `SIGPIPE` handler to default before flushing the standard handles would be helpful.
